deep_qa/data/data_generator.py

```diff
@@ -86,45 +112,63 @@ def create_generator(self, dataset: IndexedDataset):
         """
         if self.dynamic_padding:
             dataset.sort_by_padding(self.text_trainer.get_instance_sorting_keys())
-        instances = dataset.instances
-        if self.adaptive_batch_sizes:
-            grouped_instances = self.__adaptive_grouping(instances)
-        else:
-            grouped_instances = group_by_count(instances, self.text_trainer.batch_size, None)
-            grouped_instances[-1] = [instance for instance in grouped_instances[-1] if instance is not None]
+        grouped_instances = self.__create_batches(dataset.instances)
         self.last_num_batches = len(grouped_instances)
         def generator():
             while True:
-                if self.biggest_batch_first:
-                    # We'll actually pop the last _two_ batches, because the last one might not
-                    # be full.
-                    last_batch = grouped_instances.pop()
-                    penultimate_batch = grouped_instances.pop()
-                    random.shuffle(grouped_instances)
-                    grouped_instances.insert(0, penultimate_batch)
-                    grouped_instances.insert(0, last_batch)
+                if self.repermute_every_epoch:
+                    groups = self.__create_batches(dataset.instances)
                 else:
-                    random.shuffle(grouped_instances)
-                for group in grouped_instances:
-                    dataset = IndexedDataset(group)
-                    dataset.pad_instances(self.text_trainer.get_padding_lengths(), verbose=False)
-                    yield dataset.as_training_data()
+                    groups = grouped_instances
+                for group in groups:
+                    batch = IndexedDataset(group)
+                    batch.pad_instances(self.text_trainer.get_padding_lengths(), verbose=False)
+                    yield batch.as_training_data()
         return generator()
 
+    def __create_batches(self, instances: List[IndexedInstance]) -> List[List[IndexedInstance]]:
+        for i in range(1, len(instances)):
+            if random.random() < self.permute_probability:
```

Comment:
Yeah, that's what would happen if the probability is 1.  I was just trying to come up with a simple way to get a slight permutation of the sorted ordering.  Can you think of a better way to do it?